ESP32 Support, General refactoring for architecture-independant code, Zero position write support.

paucarre opened the pull request:
The library doesn't work using  ESP32 microcontrollers, the pull request resolves the issue. The reason is pretty simple.  For fast microcontrollers it seems when sending a transfer, it's necessary to wait for the sensor to have data to be received.

I found several other problems, such as the use of *word*  or *int* types which are not architecture-independant. I also seized the opportunity to remove a few class variables not used, and modernize the code a bit (constructor assignment, constants, variable declaration...)

Furthermore, I simplified the *write* method, removed all the code that splits and merges  16 bits variables into 8 bits. Same shall be done for *write*, which I will probably do later on.

Finally I added methods to get the rotation in degrees and radians.

Thanks for the library by the way!

... two more things:

The code is tested with Arduino UNO and ESP32. It seems to work for both.
I added sketches for both and renamed them so that they fit specifically the microcontroller type.

paucarre committed:
Support for esp32. Architecture-independant types. Code style clean-up (removed unused class variables, proper C++ initialization, simplified write...). Added methods to get rotation in degrees and radians. Updated sketches adding esp32 example.

paucarre committed:
disable debug

paucarre committed:
support full backwards compatibility

paucarre committed:
updated revision number

paucarre committed:
typo

paucarre committed:
reverted unnecessary change

paucarre committed:
Updated doc. Simplified radians conversion

benhowes commented:
@paucarre, this is fantastic! Most of the team is away at the moment, including me, however I will make sure that this gets tested and merged over the next couple of weeks. 

Thanks ever so much for this and I think it's a really good contribution to the community

paucarre commented:
No problem, take as much time as you need, just wanted to know if there were people notified of PR. So far it's working well for my purpose https://drive.google.com/file/d/1DPY4aPaDYzX_SdBavivNrGoniEyXbJqI/view

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -2,25 +2,39 @@
+
+static const float AS5048A_MAX_VALUE = 8191.0;
+static const float AS5048A_TWICE_MAX_VALUE = AS5048A_MAX_VALUE * 2.0;
+static const float AS5048A_PI  = 3.14159265358979323846;
```
This variable is not necessary. You can use `PI` instead.
An example program would be:

```
void setup() {
  Serial.begin(9600);
  Serial.print(PI,20);
}

void loop() {}
```

And you will obtain as result `3.14159274101257324218`.

You can use `M_PI` but it's not very precise.

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -2,25 +2,39 @@
 
 /**
- * Constructor
+ * Constructor usign response delay (ESP32 and similars)
```
using

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -68,42 +81,62 @@ byte AS5048A::spiCalcEvenParity(word value){
-	data = AS5048A::getRawRotation();
-	rotation = (int)data - (int)position;
+	uint16_t data = AS5048A::getRawRotation();
+	int32_t rotation = (int32_t)data - (int32_t)position;
```
Consider using C++ casting `static_cast<int32_t>(var)`.

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -68,42 +81,62 @@ byte AS5048A::spiCalcEvenParity(word value){
+	uint16_t data = AS5048A::getRawRotation();
+	int32_t rotation = (int32_t)data - (int32_t)position;
 	if(rotation > 8191) rotation = -((0x3FFF)-rotation); //more than -180
 	//if(rotation < -0x1FFF) rotation = rotation+0x3FFF;
```
I would remove commented code.

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -68,42 +81,62 @@ byte AS5048A::spiCalcEvenParity(word value){
+
+float AS5048A::getRotationInRadians(){
+	int32_t rotation = getRotation();
+	float degrees = AS5048A_PI * (rotation + AS5048A_MAX_VALUE) / AS5048A_MAX_VALUE;
```
`float radians`

eborghi10 replied:
You can also remove the variable.

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -68,42 +81,62 @@ byte AS5048A::spiCalcEvenParity(word value){
 /**
  * Print the diagnostic register of the sensor
  */
 void AS5048A::printState(){
```
This method should be wrapped with the `if(debug)` clause.

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -114,54 +147,50 @@ void AS5048A::printState(){
-	return (byte) data & 0xFF;
+uint8_t AS5048A::getGain(){
+	uint16_t data = AS5048A::getState();
+	return (uint8_t) data & 0xFF;
```
`static_cast<uint8_t>()`

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -175,28 +204,28 @@ word AS5048A::read(word registerAddress){
 	digitalWrite(_cs, LOW);
-	SPI.transfer(left_byte);
-	SPI.transfer(right_byte);
+	SPI.transfer16(command);
```
Good idea!

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -175,28 +204,28 @@ word AS5048A::read(word registerAddress){
 	digitalWrite(_cs,HIGH);
 
+	if(response_delay_millis > 0) {
+		delay(response_delay_millis);
```
You can use `delay()` without the `if()` if you implement the optional argument in the constructor.

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -245,12 +276,12 @@ word AS5048A::write(word registerAddress, word data) {
 	SPI.transfer(left_byte);
```
Why didn't you apply the same logic as in `read()`?

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -264,7 +295,11 @@ word AS5048A::write(word registerAddress, word data) {
+
 	//Send a NOP to get the new data in the register
 	digitalWrite(_cs, LOW);
 	left_byte =-SPI.transfer(0x00);
```
Is this the reason why you didn't apply the 16-bit logic?

eborghi10 commented on lib/AS5048A/AS5048A.h:
```diff
@@ -37,38 +34,38 @@ class AS5048A{
 	/**
 	 * Get the rotation of the sensor relative to the zero position.
 	 *
 	 * @return {int} between -2^13 and 2^13
```
`{int32_t}`

eborghi10 commented on src/sketch_esp32_nodemcu.ino:
```diff
@@ -0,0 +1,23 @@
+	Serial.print("Got rotation of: 0x");
+	Serial.println(val, HEX);
+	Serial.print("State: ");
+	angleSensor.printState();
```
If you don't set `Serial.begin()`, this line will crash the app.

eborghi10 commented on lib/AS5048A/AS5048A.cpp:
```diff
@@ -68,42 +81,62 @@ byte AS5048A::spiCalcEvenParity(word value){
-	rotation = (int)data - (int)position;
+	uint16_t data = AS5048A::getRawRotation();
+	int32_t rotation = (int32_t)data - (int32_t)position;
 	if(rotation > 8191) rotation = -((0x3FFF)-rotation); //more than -180
```
`if (rotation > AS5048A_MAX_VALUE)`

eborghi10 reviewed (changes requested):
Thank you so much for your work.
I left a couple of comments, PTAL.

BTW, I tested on [my ESP32](https://www.aliexpress.com/item/Goouuu-esp32-development-board-wireless-WiFi-Bluetooth-2-and-1-dual-core-CPU-core-board-ESP/32799809384.html) and it's not working.

Commit:
By default disable sensor. It will be enabled whenever it needs to read

paucarre committed:
Support to write zero position (#1)

Added support to write the zero position using SPI
General refactoring of the library speeding up parity bit computation as well as reducing the total number of lines increasing code-clarity

paucarre committed:
Added angle inversion option and improved documentation
